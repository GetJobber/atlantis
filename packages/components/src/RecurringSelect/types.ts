export enum WeekDay {
  Sunday,
  Monday,
  Tuesday,
  Wednesday,
  Thursday,
  Friday,
  Saturday,
}
export type DayOfMonth =
  | 1
  | 2
  | 3
  | 4
  | 5
  | 6
  | 7
  | 8
  | 9
  | 10
  | 11
  | 12
  | 13
  | 14
  | 15
  | 16
  | 17
  | 18
  | 19
  | 20
  | 21
  | 22
  | 23
  | 24
  | 25
  | 26
  | 27
  | 28
  | 29
  | 30
  | 31
  | "LAST";

export enum DurationPeriod {
  Day = "Day",
  Week = "Week",
  DayOfMonth = "DayOfMonth",
  WeekDayOfMonth = "WeekDayOfMonth",
  Year = "Year",
  Visit = "Visit",
}
export interface RecurrenceRuleDay {
  type: DurationPeriod.Day;
}

export interface RecurrenceRuleWeek {
  type: DurationPeriod.Week;
  weekDays: Set<WeekDay>;
}

export interface RecurrenceRuleYear {
  type: DurationPeriod.Year;
}

export interface RecurrenceRuleDayOfMonth {
  type: DurationPeriod.DayOfMonth;
  date: Set<DayOfMonth>;
}

export interface RecurrenceRuleWeekDayOfMonth {
  type: DurationPeriod.WeekDayOfMonth;
  dayOfWeek: [Set<WeekDay>, Set<WeekDay>, Set<WeekDay>, Set<WeekDay>];
}

export type RecurrenceRule = {
  interval: number;
} & (
  | RecurrenceRuleDay
  | RecurrenceRuleWeek
  | RecurrenceRuleYear
  | RecurrenceRuleDayOfMonth
  | RecurrenceRuleWeekDayOfMonth
);
export interface InitialStateMap {
  [DurationPeriod.Day]: {
    type: DurationPeriod.Day;
  };
  [DurationPeriod.Week]: {
    type: DurationPeriod.Week;
    weekDays: Set<WeekDay>;
  };
  [DurationPeriod.DayOfMonth]: {
    type: DurationPeriod.DayOfMonth;
    date: Set<DayOfMonth>;
  };
  [DurationPeriod.WeekDayOfMonth]: {
    type: DurationPeriod.WeekDayOfMonth;
    dayOfWeek: [Set<WeekDay>, Set<WeekDay>, Set<WeekDay>, Set<WeekDay>];
  };
  [DurationPeriod.Year]: {
    type: DurationPeriod.Year;
  };
}

export const typeInitialStateMap: InitialStateMap = {
  [DurationPeriod.Day]: {
    type: DurationPeriod.Day,
  },
  [DurationPeriod.Week]: {
    type: DurationPeriod.Week,
    weekDays: new Set<WeekDay>(),
  },
  [DurationPeriod.DayOfMonth]: {
    type: DurationPeriod.DayOfMonth,
    date: new Set<DayOfMonth>(),
  },
  [DurationPeriod.WeekDayOfMonth]: {
    type: DurationPeriod.WeekDayOfMonth,
    dayOfWeek: [
      new Set<WeekDay>(),
      new Set<WeekDay>(),
      new Set<WeekDay>(),
      new Set<WeekDay>(),
    ],
  },
  [DurationPeriod.Year]: {
    type: DurationPeriod.Year,
  },
};

export const isMonthly = (
  rule: RecurrenceRule,
): rule is Extract<
  RecurrenceRule,
  { type: DurationPeriod.WeekDayOfMonth | DurationPeriod.DayOfMonth }
> =>
  rule.type === DurationPeriod.WeekDayOfMonth ||
  rule.type === DurationPeriod.DayOfMonth;
