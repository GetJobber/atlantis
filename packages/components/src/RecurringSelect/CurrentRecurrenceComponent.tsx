import React from "react";
import {
  DayOfMonthSelect,
  MonthlyDayOfWeekSelect,
  WeeklySelect,
} from "./components";
import {
  DayOfMonth,
  DurationPeriod,
  RecurrenceRule,
  RecurrenceRuleWeekDayOfMonth,
  WeekDay,
} from "./types";

interface RecurringSelectProps {
  readonly disabled?: boolean;
  readonly recurrenceRule: RecurrenceRule;
  onChange(newRecurrence: RecurrenceRule): void;
}

export function CurrentRecurrenceComponent({
  recurrenceRule,
  disabled = false,
  onChange,
}: RecurringSelectProps) {
  switch (recurrenceRule.type) {
    case DurationPeriod.Week: {
      const onChangeWeekDays = (next: Set<WeekDay>): void => {
        onChange({
          ...recurrenceRule,
          weekDays: next,
        });
      };

      return (
        <WeeklySelect
          disabled={disabled}
          selectedDays={recurrenceRule.weekDays}
          onChange={onChangeWeekDays}
        />
      );
    }

    case DurationPeriod.WeekDayOfMonth: {
      const onChangeWeekDayOfMonth = (
        next: RecurrenceRuleWeekDayOfMonth["dayOfWeek"],
      ): void => {
        onChange({
          ...recurrenceRule,
          dayOfWeek: next,
        });
      };

      return (
        <MonthlyDayOfWeekSelect
          disabled={disabled}
          onChange={onChangeWeekDayOfMonth}
          selectedWeeks={recurrenceRule.dayOfWeek}
        />
      );
    }

    case DurationPeriod.DayOfMonth: {
      const onChangeDayOfMonth = (next: Set<DayOfMonth>): void => {
        onChange({
          ...recurrenceRule,
          date: next,
        });
      };

      return (
        <DayOfMonthSelect
          disabled={disabled}
          selectedDays={recurrenceRule.date}
          onChange={onChangeDayOfMonth}
        />
      );
    }

    default:
      return <></>;
  }
}
