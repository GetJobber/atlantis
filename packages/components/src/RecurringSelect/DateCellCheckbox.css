.container {
  display: block;
}

.checkboxWrapper {
  position: relative;
}

/* kept present but hidden for accessibility */
.container input[type="checkbox"] {
  clip: rect(1px, 1px, 1px, 1px);
  clip-path: inset(50%);
  position: absolute;
  width: 1px;
  height: 1px;
  margin: -1px;
  padding: 0;
  overflow: hidden;
}

.dayCheckbox {
  display: flex;
  justify-content: center;
  align-items: center;
  padding: var(--space-smaller);
  border-radius: var(--radius-large);
  text-align: center;
  user-select: none;
  background: var(--color-surface);
  transition: var(--timing-base);
}

@supports (aspect-ratio: 1/1) {
  .dayCheckbox {
    /*
     * disabling property-no-unknown here as this is a real property
     * that stylelint just doesn't know about. We may be able to remove
     * this when we update our stylelint to not care about IE11.
     */
    /* stylelint-disable-next-line property-no-unknown */
    aspect-ratio: 1/1;
  }
}

.dayCheckbox:hover {
  background: var(--color-surface--hover);
}

.container input[type="checkbox"]:focus + .dayCheckbox {
  box-shadow: var(--shadow-focus);
  background: var(--color-surface--hover);
}

.container input[type="checkbox"]:checked + .dayCheckbox {
  color: var(--color-text--reverse);
  background: var(--color-interactive);
}

.container input[type="checkbox"]:disabled + .dayCheckbox {
  color: var(--color-disabled);
  background: var(--color-surface--background);
  cursor: default;
}

.container input[type="checkbox"]:enabled + .dayCheckbox {
  cursor: pointer;
}

.container input[type="checkbox"]:checked:focus + .dayCheckbox {
  background: var(--color-interactive--hover);
}

.container input[type="checkbox"]:checked + .dayCheckbox:hover {
  background: var(--color-interactive--hover);
}

.container input[type="checkbox"]:disabled:checked + .dayCheckbox {
  color: var(--color-disabled--secondary);
  background: var(--color-disabled);
}
