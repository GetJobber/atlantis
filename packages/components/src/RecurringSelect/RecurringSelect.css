.container {
  display: flex;
  align-items: center;
  gap: var(--space-base);
}

.fullWidthWrapper {
  width: 100%;
}
