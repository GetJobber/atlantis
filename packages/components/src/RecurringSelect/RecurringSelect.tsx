import React, { ReactNode } from "react";
import classNames from "classnames";
import { DayOfMonthSelect } from "./components/DayOfMonthSelect/DayOfMonthSelect";
import styles from "./RecurringSelect.css";
import { WeeklySelect } from "./components/WeeklySelect/WeeklySelect";
import { MonthlySelect } from "./components/MonthlySelect/MonthlySelect";
import { MonthlyDayOfWeekSelect } from "./components/MonthlyDayOfWeekSelect/MonthlyDayOfWeekSelect";
import {
  DayOfMonth,
  DurationPeriod,
  RecurrenceRule,
  RecurrenceRuleWeekDayOfMonth,
  ScheduleEnd,
  SchedulerState,
  WeekDay,
  typeInitialStateMap,
} from "./types";
import { Option, Select } from "../Select";
import { InputNumber } from "../InputNumber";
import { InputGroup } from "../InputGroup";
import { Content } from "../Content";

interface RecurringSelectProps {
  disabled: boolean;
  recurrenceRule: RecurrenceRule;
  recurrenceEnds: ScheduleEnd;
  onChange(newPartialState: Partial<SchedulerState>): void;
}

export function RecurringSelect({
  recurrenceRule,
  recurrenceEnds,
  disabled,
  onChange,
}: RecurringSelectProps) {
  const onRecurrenceChange = (newRecurrence: {
    ends: ScheduleEnd;
    rule: RecurrenceRule;
  }) =>
    onChange({
      recurrence: {
        ...newRecurrence,
      },
    });
  const currentRecurrenceComponent = getCurrentComponent(
    recurrenceRule,
    recurrenceEnds,
    onRecurrenceChange,
    disabled,
  );
  const monthlySelectOption =
    recurrenceRule.type === DurationPeriod.WeekDayOfMonth ? (
      <Option value={DurationPeriod.WeekDayOfMonth}>Month(s)</Option>
    ) : (
      <Option value={DurationPeriod.DayOfMonth}>Month(s)</Option>
    );
  let extraFrequencyDescriptor = <></>;
  let monthlySelectEl = <></>;

  if (
    recurrenceRule.type === DurationPeriod.WeekDayOfMonth ||
    recurrenceRule.type === DurationPeriod.DayOfMonth ||
    recurrenceRule.type === DurationPeriod.Week
  ) {
    extraFrequencyDescriptor = <span className={styles.modifierText}>on</span>;
  }

  if (
    recurrenceRule.type === DurationPeriod.WeekDayOfMonth ||
    recurrenceRule.type === DurationPeriod.DayOfMonth
  ) {
    // we must dynamically populate the select option based on which is selected
    // because there is no single "month" option, it must always be one of these two
    monthlySelectEl = (
      <MonthlySelect
        disabled={disabled}
        onChange={onChangeType}
        selectedMonthOption={recurrenceRule.type}
      />
    );
  }

  return (
    <Content>
      <div className={styles.container}>
        <span
          className={classNames(
            styles.intervalText,
            disabled ? styles.disabled : "",
          )}
        >
          Every
        </span>
        <div className={styles.fullWidthWrapper}>
          <InputGroup flowDirection="horizontal">
            <InputNumber
              disabled={disabled}
              name="schedule-recurrence-interval"
              value={recurrenceRule.interval}
              min={1}
              maxLength={3}
              onChange={onChangeInterval}
            />

            <Select
              disabled={disabled}
              value={recurrenceRule.type}
              onChange={onChangeType}
              name="schedule-recurrence-type"
            >
              <Option value={DurationPeriod.Day}>Day(s)</Option>
              <Option value={DurationPeriod.Week}>Week(s)</Option>
              {monthlySelectOption}
              <Option value={DurationPeriod.Year}>Year(s)</Option>
            </Select>
          </InputGroup>
        </div>
        {extraFrequencyDescriptor}
      </div>
      {monthlySelectEl}
      {currentRecurrenceComponent}
    </Content>
  );

  function onChangeInterval(interval: number) {
    onRecurrenceChange({
      rule: {
        ...recurrenceRule,
        interval,
      },
      ends: recurrenceEnds,
    });
  }

  function onChangeType(
    type:
      | DurationPeriod.Day
      | DurationPeriod.Week
      | DurationPeriod.DayOfMonth
      | DurationPeriod.WeekDayOfMonth
      | DurationPeriod.Year,
  ) {
    onRecurrenceChange({
      rule: {
        interval: recurrenceRule.interval,
        ...typeInitialStateMap[type],
      },
      ends: recurrenceEnds,
    });
  }
}

function getCurrentComponent(
  recurrenceRule: RecurrenceRule,
  recurrenceEnds: ScheduleEnd,
  callback: (newRecurrence: {
    rule: RecurrenceRule;
    ends: ScheduleEnd;
  }) => void,
  disabled: boolean,
): ReactNode {
  if (recurrenceRule.type === DurationPeriod.Week) {
    const onChangeWeekDays = (next: Set<WeekDay>): void => {
      callback({
        rule: {
          ...recurrenceRule,
          weekDays: next,
        },
        ends: recurrenceEnds,
      });
    };

    return (
      <WeeklySelect
        disabled={disabled}
        selectedDays={recurrenceRule.weekDays}
        onChange={onChangeWeekDays}
      />
    );
  }

  if (recurrenceRule.type === DurationPeriod.WeekDayOfMonth) {
    const onChangeWeekDayOfMonth = (
      next: RecurrenceRuleWeekDayOfMonth["dayOfWeek"],
    ): void => {
      callback({
        rule: {
          ...recurrenceRule,
          dayOfWeek: next,
        },
        ends: recurrenceEnds,
      });
    };

    return (
      <MonthlyDayOfWeekSelect
        disabled={disabled}
        onChange={onChangeWeekDayOfMonth}
        selectedWeeks={recurrenceRule.dayOfWeek}
      />
    );
  }

  if (recurrenceRule.type === DurationPeriod.DayOfMonth) {
    const onChangeDayOfMonth = (next: Set<DayOfMonth>): void => {
      callback({
        rule: {
          ...recurrenceRule,
          date: next,
        },
        ends: recurrenceEnds,
      });
    };
    return (
      <>
        {recurrenceRule.type === DurationPeriod.DayOfMonth && (
          <DayOfMonthSelect
            disabled={disabled}
            selectedDays={recurrenceRule.date}
            onChange={onChangeDayOfMonth}
          />
        )}
      </>
    );
  }
}
