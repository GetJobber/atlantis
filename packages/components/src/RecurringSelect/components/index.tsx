export { WeeklySelect } from "./WeeklySelect/WeeklySelect";
