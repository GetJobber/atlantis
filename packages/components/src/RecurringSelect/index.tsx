export { RecurringSelect } from "./RecurringSelect";
export { DayOfMonth, DurationPeriod, RecurrenceRule, WeekDay } from "./types";
