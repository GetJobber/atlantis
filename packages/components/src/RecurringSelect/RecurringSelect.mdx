---
name: RecurringSelect
menu: Components
route: /components/recuringselect
showDirectoryLink: true
---

import { Playground, Props } from "docz";
import { ComponentStatus } from "@jobber/docx";
import { RecurringSelect } from "./RecurringSelect";
import { Content } from "@jobber/components/Content";
import { Text } from "@jobber/components/Text";
import { Checkbox } from "@jobber/components/Checkbox";
import { Divider } from "@jobber/components/Divider";
import { Avatar } from "@jobber/components/Avatar";
import { useState } from "react";
import { DurationPeriod, SchedulerState } from "./types";
import { useScheduler } from "./useScheduler";

# RecurringSelect

<ComponentStatus stage="ready" responsive="yes" accessible="yes" />

RecurringSelect, fill me out & rename me please

```ts
import { RecurringSelect } from "@jobber/components/RecurringSelect";
```

<Playground>
  {() => {
    const [schedulerState, setSchedulerState] = useScheduler({
      recurrence: {
        rule: { interval: 1, type: DurationPeriod.Day },
        ends: {
          type: "date",
          date: new Date(),
          numOfPeriods: 1,
          durationPeriod: DurationPeriod.Day,
        },
      },
    });
    const onChangeSchedule = (partialNextSchedule) =>
      setSchedulerState({ ...schedulerState, ...partialNextSchedule });
    return(
      <RecurringSelect
        recurrenceRule={schedulerState.recurrence.rule}
        recurrenceEnds={schedulerState.recurrence.ends}
        disabled={false}
        onChange={onChangeSchedule}
      />);
    }}

</Playground>
