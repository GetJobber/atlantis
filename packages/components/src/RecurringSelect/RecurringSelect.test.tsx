/* eslint-disable max-statements */
/* tslint:disable no-relative-imports */
import React from "react";
import { cleanup, fireEvent, render } from "@testing-library/react";
import renderer from "react-test-renderer";
import { RecurringSelect } from "./RecurringSelect";
// eslint-disable-next-line
import { DayOfMonth, DurationPeriod, WeekDay } from "./types";

afterEach(cleanup);

let onChange: jest.Mock;

describe("RecurringSelect", () => {
  beforeEach(() => {
    onChange = jest.fn();
  });

  it("should render the recurring select component", () => {
    const tree = renderer
      .create(
        <RecurringSelect
          data-testid="boop"
          recurrenceRule={{ interval: 1, type: DurationPeriod.Day }}
          recurrenceEnds={{
            type: "date",
            date: new Date("December 17, 2021 03:24:00"),
            numOfPeriods: 1,
            durationPeriod: DurationPeriod.Day,
          }}
          onRecurrenceChange={onChange}
          disabled={false}
        />,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe("RecurringSelect with a weekly recurrenceRule", () => {
  beforeEach(() => {
    onChange = jest.fn();
  });

  it("should render the days of the WeeklySelect component", () => {
    const { queryAllByText, queryByLabelText } = render(
      <RecurringSelect
        data-testid="boop"
        recurrenceRule={{
          interval: 1,
          type: DurationPeriod.Week,
          weekDays: new Set(),
        }}
        recurrenceEnds={{
          type: "date",
          date: new Date("December 17, 2021 03:24:00"),
          numOfPeriods: 1,
          durationPeriod: DurationPeriod.Day,
        }}
        onRecurrenceChange={onChange}
        disabled={false}
      />,
    );

    expect(queryAllByText(/S/)).toBeTruthy();
    expect(queryAllByText(/T/)).toBeTruthy();
    expect(queryByLabelText(/F/)).toBeTruthy();
    expect(queryByLabelText(/W/)).toBeTruthy();
    expect(queryByLabelText(/M/)).toBeTruthy();
  });

  it("should call the onChange when checkbox is clicked", () => {
    const { getByLabelText } = render(
      <RecurringSelect
        data-testid="boop"
        recurrenceRule={{
          interval: 1,
          type: DurationPeriod.Week,
          weekDays: new Set(),
        }}
        recurrenceEnds={{
          type: "date",
          date: new Date("December 17, 2021 03:24:00"),
          numOfPeriods: 1,
          durationPeriod: DurationPeriod.Day,
        }}
        onRecurrenceChange={onChange}
        disabled={false}
      />,
    );

    fireEvent.click(getByLabelText(/F/i));
    expect(onChange).toHaveBeenCalled();
  });
});

describe("RecurringSelect with an existing week day selected", () => {
  beforeEach(() => {
    onChange = jest.fn();
  });

  it("should call the onChange when checkbox is clicked", () => {
    const newWeekDaySet = new Set<WeekDay>();
    newWeekDaySet.add(WeekDay.Friday);

    const { getByLabelText } = render(
      <RecurringSelect
        data-testid="boop"
        recurrenceRule={{
          interval: 1,
          type: DurationPeriod.Week,
          weekDays: newWeekDaySet,
        }}
        recurrenceEnds={{
          type: "date",
          date: new Date("December 17, 2021 03:24:00"),
          numOfPeriods: 1,
          durationPeriod: DurationPeriod.Day,
        }}
        onRecurrenceChange={onChange}
        disabled={false}
      />,
    );

    fireEvent.click(getByLabelText(/F/i));
    expect(onChange).toHaveBeenCalled();
  });
});

describe("RecurringSelect with a daily recurrence", () => {
  beforeEach(() => {
    onChange = jest.fn();
  });

  it("should call the onChange when interval and type are changed", () => {
    const { container } = render(
      <RecurringSelect
        data-testid="boop"
        recurrenceRule={{
          interval: 1,
          type: DurationPeriod.Day,
        }}
        recurrenceEnds={{
          type: "date",
          date: new Date("December 17, 2021 03:24:00"),
          numOfPeriods: 1,
          durationPeriod: DurationPeriod.Day,
        }}
        onRecurrenceChange={onChange}
        disabled={false}
      />,
    );

    const intervalEl = container.querySelector(
      'input[name="schedule-recurrence-interval"]',
    );
    expect(intervalEl).not.toBeNull();
    expect(intervalEl instanceof HTMLInputElement).toBe(true);

    if (intervalEl) {
      fireEvent.change(intervalEl, { target: { value: 3 } });

      expect(onChange).toHaveBeenCalledTimes(1);

      const typeEl = container.querySelector(
        'select[name="schedule-recurrence-type"]',
      );
      expect(typeEl).not.toBeNull();
      expect(typeEl instanceof HTMLSelectElement).toBe(true);

      if (typeEl) {
        fireEvent.change(typeEl, { target: { value: "Year" } });
      }

      expect(onChange).toHaveBeenCalledTimes(2);
    }
  });
});

describe("RecurringSelect with a day of month recurrenceRule", () => {
  beforeEach(() => {
    onChange = jest.fn();
  });

  it("should render the days of the day of month component", () => {
    const { queryAllByText, queryByLabelText } = render(
      <RecurringSelect
        data-testid="boop"
        recurrenceRule={{
          interval: 1,
          type: DurationPeriod.DayOfMonth,
          date: new Set(),
        }}
        recurrenceEnds={{
          type: "date",
          date: new Date("December 17, 2021 03:24:00"),
          numOfPeriods: 1,
          durationPeriod: DurationPeriod.Day,
        }}
        onRecurrenceChange={onChange}
        disabled={false}
      />,
    );

    expect(queryAllByText("1")).toBeTruthy();
    expect(queryAllByText("2")).toBeTruthy();
    expect(queryByLabelText("3")).toBeTruthy();
    expect(queryByLabelText("4")).toBeTruthy();
    expect(queryByLabelText("5")).toBeTruthy();
  });

  it("should call the onChange when checkbox is clicked", () => {
    const newDayOfMonthSet = new Set<DayOfMonth>();
    const { getByLabelText } = render(
      <RecurringSelect
        data-testid="boop"
        recurrenceRule={{
          interval: 1,
          type: DurationPeriod.DayOfMonth,
          date: newDayOfMonthSet,
        }}
        recurrenceEnds={{
          type: "date",
          date: new Date("December 17, 2021 03:24:00"),
          numOfPeriods: 1,
          durationPeriod: DurationPeriod.Day,
        }}
        onRecurrenceChange={onChange}
        disabled={false}
      />,
    );

    fireEvent.click(getByLabelText("1"));
    fireEvent.click(getByLabelText("2"));
    expect(onChange).toHaveBeenCalled();
  });

  it("should call the onChange when checkbox is unchecked", () => {
    const newDayOfMonthSet = new Set<DayOfMonth>();
    newDayOfMonthSet.add(1);
    const { getByLabelText } = render(
      <RecurringSelect
        data-testid="boop"
        recurrenceRule={{
          interval: 1,
          type: DurationPeriod.DayOfMonth,
          date: newDayOfMonthSet,
        }}
        recurrenceEnds={{
          type: "date",
          date: new Date("December 17, 2021 03:24:00"),
          numOfPeriods: 1,
          durationPeriod: DurationPeriod.Day,
        }}
        onRecurrenceChange={onChange}
        disabled={false}
      />,
    );

    fireEvent.click(getByLabelText("1"));
    expect(onChange).toHaveBeenCalled();
  });

  it("should render when last day selected", () => {
    const newDayOfMonthSet = new Set<DayOfMonth>();
    newDayOfMonthSet.add("LAST");
    const { getByLabelText } = render(
      <RecurringSelect
        data-testid="boop"
        recurrenceRule={{
          interval: 1,
          type: DurationPeriod.DayOfMonth,
          date: newDayOfMonthSet,
        }}
        recurrenceEnds={{
          type: "date",
          date: new Date("December 17, 2021 03:24:00"),
          numOfPeriods: 1,
          durationPeriod: DurationPeriod.Day,
        }}
        onRecurrenceChange={onChange}
        disabled={false}
      />,
    );

    fireEvent.click(getByLabelText(/Last day/i));
    expect(onChange).toHaveBeenCalled();
  });
});
