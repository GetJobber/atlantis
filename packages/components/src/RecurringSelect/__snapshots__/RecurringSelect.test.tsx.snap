// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecurringSelect should render the recurring select component 1`] = `
<div
  className="_1TB2WRsjEoAUFq0FMfwdQ3 _3XB3f5356G3URhbYQmxO0U"
>
  <div
    className="container"
  >
    <span
      className="intervalText"
    >
      Every
    </span>
    <div
      className="fullWidthWrapper"
    >
      <div
        className="_3kywSyvvwSdmNSTOqX-AQA _26qQBed5Rzu9i2ZIDEaESr"
      >
        <div
          className="_2bOj_Hpd2VvA3mi9t4RcZO _1tABQjw4-IkgUE3cbBzs7r"
          style={
            Object {
              "--formField-maxLength": 3,
            }
          }
        >
          <div
            className="_1WnxEaBZY0zLsD8oKwh20b"
          >
            <label
              className="ktPFw5_r1v4MB-Ts5rPTG"
              htmlFor="123e4567-e89b-12d3-a456-426655440001"
            />
            <input
              className="_3koQsrzZZXPVxXmmESjSG_"
              disabled={false}
              id="123e4567-e89b-12d3-a456-426655440001"
              maxLength={3}
              min={1}
              name="schedule-recurrence-interval"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              type="number"
              value={1}
            />
          </div>
        </div>
        <div
          className="_2bOj_Hpd2VvA3mi9t4RcZO Ntcl388Qhhu4uLJc2d1Ll"
          style={
            Object {
              "--formField-maxLength": undefined,
            }
          }
        >
          <div
            className="_1WnxEaBZY0zLsD8oKwh20b"
          >
            <label
              className="ktPFw5_r1v4MB-Ts5rPTG"
              htmlFor="123e4567-e89b-12d3-a456-426655440003"
            />
            <select
              className="_3koQsrzZZXPVxXmmESjSG_"
              disabled={false}
              id="123e4567-e89b-12d3-a456-426655440003"
              name="schedule-recurrence-type"
              onChange={[Function]}
              value="Day"
            >
              <option
                value="Day"
              >
                Day(s)
              </option>
              <option
                value="Week"
              >
                Week(s)
              </option>
              <option
                value="DayOfMonth"
              >
                Month(s)
              </option>
              <option
                value="Year"
              >
                Year(s)
              </option>
            </select>
            <span
              className="_1yJbSG5N0hfPjPkrih4Q99"
            >
              <svg
                className="ZDlJU6tECg7ouG-b27gya _3ctOeOMP7zLhU37n25xoZC"
                data-testid="arrowDown"
                viewBox="0 0 1024 1024"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  className=""
                  d="M328.677 353.767c-8.333-8.309-19.249-12.453-30.16-12.433-10.876 0.019-21.745 4.175-30.042 12.466s-12.456 19.153-12.475 30.021c-0.019 10.918 4.139 21.842 12.475 30.172l213.334 213.51c8.337 8.333 19.264 12.497 30.191 12.497s21.854-4.164 30.191-12.497l213.333-213.33c8.337-8.33 12.497-19.254 12.476-30.172-0.021-10.868-4.177-21.729-12.476-30.021-8.294-8.291-19.166-12.447-30.042-12.466-10.91-0.019-21.828 4.125-30.161 12.433l-183.322 182.979-183.323-183.159z"
                />
              </svg>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
